Added support for an indicator constraint to QuadraticProgram

a-matsuo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
resolve #13
The PR adds support for an [indicator constraint](http://ibmdecisionoptimization.github.io/docplex-doc/mp/docplex.mp.model.html#docplex.mp.model.Model.add_indicator) to `QuadraticProgram`. 

### Details and comments
An indicator constraint control  the satisfaction of a linear constraint by the value of a binary variable. If the binary variable equals the active value, then the constraint must be satisfied, but otherwise the constraint might or might not be satisfied.
e.g. `x = 1 -> y + z = 1`
- [x] to_docplex
- [x] from_docplex
- [x] converter
- [x] unittests
  - [x] indicator to inequality converter
  - [x] indicator constraint

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1025,6 +1144,40 @@ def from_docplex(self, model: Model) -> None:
                 raise QiskitOptimizationError("Unsupported constraint sense: {}".format(constraint))
 
+        # get indicator constraints
+        for constraint in model.iter_logical_constraints():
```
Isn't it `iter_indicator_constraints`? https://ibmdecisionoptimization.github.io/docplex-doc/mp/docplex.mp.model.html?highlight=iter_#docplex.mp.model.Model.iter_indicator_constraints

a-matsuo replied:
`iter_logical_constraints()` works since it's the super class of `IndicatorConstraint`, but I think `iter_indicator_constraint` is more specific and more appropriate. I fixed it.

t-imamichi commented on qiskit_optimization/problems/indicator_constraint.py:
```diff
@@ -0,0 +1,131 @@
+        Returns:
+            The left-hand-side of the constraint given the variable values.
+        """
+        return self.linear.evaluate(x)
```
I think you need to check the active value here.

a-matsuo replied:
I added `evaluate_indicator` to check whether an indicator binary variable equals to `active_value` or not. Feasibility of indicator constraints is checked in `QuadraticProgram.get_feasibility_info()`

ikkoham commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,179 @@
+            QiskitOptimizationError: If an unsupported mode is selected.
+            QiskitOptimizationError: If an unsupported sense is specified.
+        """
+        self._src = copy.deepcopy(problem)
```
Just a question. Why is this `deepcopy` needed?

t-imamichi replied:
It used to be necessary. Matsuo-san can remove it.

a-matsuo commented:
I think the PR is ready for review. Could you please review this?

t-imamichi commented:
Thanks. Could you write a reno?

a-matsuo commented:
Sure (I had written reno but I had forgotten to add it).

woodsp-ibm commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,175 @@
+                    name=x.name, lowerbound=x.lowerbound, upperbound=x.upperbound
+                )
+            else:
+                raise QiskitOptimizationError("Unsupported variable type {}".format(x.vartype))
```
I think generally now we have gone over to using f Strings now, since we dropped 3.5 Python support, rather than using this explicit .format() syntax. Not that its incorrect of course to do this, just that going forwards I thin it might be preferable to stick to f Strings so its all done one way.

woodsp-ibm commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -105,6 +107,9 @@ def __init__(self, name: str = "") -> None:
         self._quadratic_constraints = []  # type: List[QuadraticConstraint]
         self._quadratic_constraints_index = {}  # type: Dict[str, int]
 
+        self._indicator_constraints = []  # type: List[IndicatorConstraint]
```
Just another note going forwards. Since we dropped Python 3.5 support, where the typehints had to be done as comments, since 3.6 in the minimum now we can do typehints directly on the members so we don't have to do it via comments anymore.

woodsp-ibm commented on test/problems/test_indicator_constraint.py:
```diff
@@ -0,0 +1,114 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+""" Test LinearConstraint """
```
```suggestion
""" Test IndicatorConstraint """
```

woodsp-ibm commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,175 @@
+        """
+        if len(x) != self._src_num_var:
+            raise QiskitOptimizationError(
+                "The number of variables in the passed result differs from "
```
Might be nice to include the different variable counts in the message.

woodsp-ibm commented on qiskit_optimization/problems/indicator_constraint.py:
```diff
@@ -0,0 +1,181 @@
+                if index == ind:
+                    val = value
+        else:
+            raise QiskitOptimizationError("Unsupported format for x.")
```
Should the message, since its more related to type, not format, of x be more like 
`f"Unsupported type '{type(x)}' for x."` so it says type and we include the type in the message

woodsp-ibm commented on releasenotes/notes/ineq_to_pen-229bc20575b2f483.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features:
```
I assume the above reno is correct - this one looks like it should be removed

woodsp-ibm commented:
Should we be adding a new tutorial, or updating an existing one, to show this new capability. For significant new features I think it would be good, since we have the tutorials now in this same repo, that they are updated/improved at the same time if the new/updated feature warrants it.

a-matsuo commented:
@woodsp-ibm Thank you for your helpful comments. I fixed the code. I haven't fixed all the f string and typehints, but I fixed them at least in this code.

t-imamichi commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,176 @@
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""The inequality to equality converter."""
```
Need to rename it

t-imamichi commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,176 @@
+from ..problems.variable import Variable
+
+
+class IndicatorToInequality(QuadraticProgramConverter):
```
I suggest `IndicatorToLinearConstraint` because this converts not only inequality constraints but also equality constraints.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -793,6 +801,97 @@ def get_num_quadratic_constraints(self) -> int:
+        name: Optional[str] = None,
+    ) -> IndicatorConstraint:
+        """Adds an indicator constraint to the quadratic program of the form:
+            z=f -> linear * x sense rhs
```
z -> binary_var
f -> active_value

t-imamichi commented on qiskit_optimization/converters/indicator_to_inequality.py:
```diff
@@ -0,0 +1,176 @@
+        new_name = indicator_const.name + self._delimiter + "indicator"
+        if sense == Constraint.Sense.LE:
+            _, lhs_ub = self._calc_linear_bounds(problem, new_linear)
+            big_m = lhs_ub - new_rhs
```
make sure that big_m is a positive value.

a-matsuo replied:
I think it does work even when big_m is a negative value in this case. When it's inactive, it becomes lhs <= the upperbounds of lhs. (When rhs is larger than the upper bound of lhs, the constraint doesn't make sense though).

t-imamichi commented:
The indicator constraints will be treated as part of a docplex translator #122
